log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/RollingFileManager.java

```diff
@@ -344,7 +344,12 @@ public synchronized void rollover() {
     }
 
     protected void createFileAfterRollover() throws IOException  {
-        setOutputStream(createOutputStream());
+        try {
```

Comment:
Why not set the file timestamp specifically?

Reply:
The code path called by line 352, `setOutputStream(createOutputStream(initialTime))`, does attempt to set the creationTime attribute explicitly.  It uses the value of initialTime (which was set previously in `rollover()`).

The 1ms sleep on line 348 was introduced as a mitigation for `RollingAppenderSizeWithTimeTest.testAppender()` failing on Windows during its rapid rollovers (confirmed to happen before any changes in this PR were applied).  A 0ms sleep was tried but that test still failed (it's possible the 1ms sleep succeeds due to a different priority system thread associated with JVM I/O needing to execute).

Does that make sense, or did I misinterpret your question ?

Reply:
That test is flaking a lot of the linux build nodes as well as when I run the test locally on a mac. Are you confident the failure is related to windows?

Reply:
I'm _not particularly confident_ the `RollingAppenderSizeWithTimeTest.testAppender()` failure is _specific to Windows_ ... I just _assumed_ that if this test was failing with Linux someone would have reported it sooner.  :wink:

If you're saying that test is "flaking a lot" on Linux and/or Mac builds too, then that probably points to the issue being more of a general JVM I/O behaviour (i.e. could impact any OS).  Maybe the 1ms sleep mitigation will be sufficient across all OS platforms ?

Reply:
If I run the test in a loop locally, the failures are resolved by adding the 1ms sleep, however I don't think we can commit that change.
Using a relatively standard logging configuration on my development machine (macbook), I can log about ~1,000,000 events to disk every second. That millisecond represents 1,000 events that aren't logged! Do we have any options that don't involve waiting?

Reply:
The 1ms sleep only occurs on log rollover ... so unless a logging configuration involves _massive numbers of rollovers_ in _very short time periods_ it shouldn't be too expensive for that minimal wait ?

A good proportion of log rollovers are done on minute/hour/day thresholds (or rollover size thresholds sufficient to hold data for at least a minute), where a 1ms sleep should be fairly negligible.  The `testAppender()` test appears to be configured to perform rollovers at a very high rate (probably a lot higher than the vast majority of timed/size rollovers).  Given that the test shows actual lost log messages (due to overwrites) it _seems_ a short sleep is _safer_ than the alternative ...

If anyone can find a way to avoid even a 1ms sleep that would be neat, but it really looks as if the JVM needs a minimal pause to dispatch something in order for the old-file-move-new-file-create to "safely complete". :disappointed:

Reply:
Filed [LOG4J2-2613](https://issues.apache.org/jira/browse/LOG4J2-2613) to track flakes.

> A good proportion of log rollovers are done on minute/hour/day thresholds (or rollover size thresholds sufficient to hold data for at least a minute), where a 1ms sleep should be fairly negligible.

In those cases with only hourly/daily thresholds, it's not the worst thing in the world, however combined with size based policies it's common to roll 1gb logs every few seconds under load.

> Given that the test shows actual lost log messages (due to overwrites) it seems a short sleep is safer than the alternative ...

I totally agree that we should prioritize fixing this, dropped messages are bad. If our only options are to sleep or to drop logs, I'd take the sleep in a heartbeat. I want to be confident that we've explored all potential avenues before applying a sleep.